README.md
# Ridibooks CRM API PHP SDK

[![Packagist](https://img.shields.io/packagist/v/ridibooks/crm.svg)](https://packagist.org/packages/ridibooks/crm)
[![travis ci](https://travis-ci.org/ridi/crm-php.svg?branch=master)](https://travis-ci.org/ridi/crm-php)

## Introduction
CRM API PHP SDK provide client of CRM API Server used in RIDI.  
This library uses [guzzlehttp/guzzle](https://github.com/guzzle/guzzle) as HTTP Client

## Install
```sh
composer require ridibooks/crm
```

## Usage
To send a message with CRM API server

```php
use Ridibooks\Crm\Client;
use Ridibooks\Crm\Notification\Identifier;
use Ridibooks\Crm\Notification\ImageType;
use Ridibooks\Crm\Notification\MessageType;
use Ridibooks\Crm\Notification\Payload\ApnsPush;
use Ridibooks\Crm\Notification\Payload\Email;
use Ridibooks\Crm\Notification\Payload\GcmPush;
use Ridibooks\Crm\Notification\Payload\NotificationCenterMessage;
use Ridibooks\Crm\Notification\Tag;

$client = new Client();
$client_with_retry_handler = Client::createWithDefaultRetry();

$identifier = new Identifier(MessageType::CMS, 'campaign-id', [Tag::CMS]);
$landing_url = 'https://ridibooks.com';
$user_id = 'u_id';
$message = '십오야 최대 4만 포인트 증정!';
$icon = 'https://active.ridibooks.com/ridibooks_noti_icon/icon_noti_15ya.png';

$expire_at = ((new \DateTime())->getTimestamp() + 3600) * 1000;
$notification_center_message = new NotificationCenterMessage(
    [$user_id],
    $identifier,
    $message,
    $icon,
    ImageType::ICON,
    $landing_url,
    $expire_at
);
$client->sendNotificationCenterMessage($notification_center_message);

$force_silent = true;
$apns_push = new ApnsPush($user_id, $message, $landing_url, $identifier, $force_silent);
$response = $client->sendApnsPush($apns_push);
echo $response->getStatusCode();

$title = '최대 4만원 포인트 혜택';
$gcm_push = new GcmPush($user_id, $title, $message, $landing_url, $icon, $identifier, $force_silent);
$response = $client->sendGcmPush($gcm_push);
echo $response->getStatusCode();

$from = '리디북스 <no-reply@ridibooks.com>';
$to = 'to@example.com';

$subject = '최대 4만원 포인트 혜택';
$html = '<h1>%recipient.greeting%</h1><p>hello world</p>';
$cc = 'cc@example.com';
$bcc = 'bcc@exmaple.com';
$recipient_variables = [
    'to@example.com' => ['greeting' => '안녕하세요'],
    'cc@example.com' => ['greeting' => 'hello'],
    'bcc@example.com' => ['greeting' => 'guten tag'],
];
$email = new Email($from, [$to], $subject, $html, $identifier, $recipient_variables, [$cc], [$bcc]);
$response = $client->sendEmail($email);
echo $response->getStatusCode();

$response = $client->getQueueStatus();
echo $response->getQueueStatus();
echo implode(',', $response->getNameOfQueues());
```

### CRM API SDK release steps

1. Create `release/{version}` branch.
1. Update `RELEASES.md`.
1. Commit & push all the changes and make a pull request.
1. After PR, tag a new release in github. This will release a new version in Packagist.


RELEASES.md
Version 3.0.0 (2020-09-14)
==========================
- NotificationMessage 에 deeplink_url 필드 제거
- NotificationCenterMessage 에 deeplink_url 필드 제거

Version 2.1.0 (2020-07-02)
==========================
- NotificationMessage 에 deeplink_url 필드 추가
- NotificationCenterMessage 에 deeplink_url 필드 추가

Version 2.0.0 (2020-01-30)
==========================
- Ridibooks\Crm\Client::getQueueStatusAsync 제거
- Ridibooks\Crm\Client::getQueueStatus 제거
- Ridibooks\Crm\Status::QueueStatusType 제거
- Ridibooks\Crm\Status\Response::QueueStatusResponse 제거

Version 1.0.0 (2018-10-29)
==========================
- PHP 7.0 지원 중단
- 재시도 핸들러가 동작하지 않는 버그 수정

Version 0.5.0 (2018-10-02)
==========================
- CRM API 서버 큐 상태 확인 API 추가

Version 0.4.4 (2018-08-13)
==========================
- 테스트 케이스 추가, 재시도 핸들러 버그 픽스

Version 0.4.3 (2018-08-01)
==========================
- 기본 재시도 핸들러를 사용할 수 있는 `createWithDefaultRetry`가 추가되었습니다.

Version 0.4.2 (2018-07-31)
==========================
- 기다무 `MessageType`, `Tag`가 추가되었습니다.

Version 0.4.1 (2018-06-27)
==========================
- 앱푸시 대량발송을 위한 `BulkPayload`가 기존 앱푸시 페이로드를 상속합니다.. 

Version 0.4.0 (2018-06-26)
==========================
- 앱푸시 대량발송을 위한 `BulkPayload`(`BulkGcmPush`, `BulkApnsPush`)가 추가되었습니다. 

Version 0.3.2 (2018-04-04)
==========================
- `Tag::createFromBid` 함수가 추가되었습니다.

Version 0.3.0 (2018-01-23)
==========================

- `Ridibooks\Crm\Client`와 `GuzzleHttp\Client`의 생성자가 서로 호환됩니다.
- APNS 무음 푸시 페이로드를 만들 수 있습니다. 알림 페이로드 생성자를 참고하세요.

Version 0.2.0 (2018-01-10)
==========================

- 알림 발송과 관련된 상수 클래스와 API 클라이언트를 제공합니다.

  ```php
  use Ridibooks\Crm\Client;
  use Ridibooks\Crm\Notification\Identifier;
  use Ridibooks\Crm\Notification\MessageType;
  use Ridibooks\Crm\Notification\Payload\ApnsPush;
  use Ridibooks\Crm\Notification\Tag;

  $identifier = new Identifier(MessageType::CMS, '5735', [Tag::CMS]);
  $payload = new ApnsPush(
      'limeburst',
      '지금 PAPER PRO 사전 예약하고 특별 혜택 받으세요!',
      'https://paper.ridibooks.com/Reservation',
      $identifier
  );

  $client = new Client();
  $res = $client->sendApnsPush($payload);
  ```
  
  [CRM API 스펙 문서](http://api.dev.ridi.io/crm/crm.html)를 참고해 주세요.

Version 0.1.0 (2017-11-13)
==========================

- 다음 두 상수 타입을 제공합니다:
  - `Ridibooks\Bi\Crm\RegularCoupon\RegularComicRentCouponConstant`
  - `Ridibooks\Bi\Notification\DarTypeConst`


src/Ridibooks/Crm/Notification/Icon.php
<?php
declare(strict_types=1);

namespace Ridibooks\Crm\Notification;

class Icon
{
    // 이하부터 상수를 tb_notification_image.image_id 와 맞춤(by Sunghoon)
    const DEFAULT = 1;
    const COUPON = 2;
    const CONGRATULATION = 3;
    const X2 = 4;
    const FIFTEEN_NIGHT = 5;
    const DISCOUNT_APPROACHING = 6;
    const DISCOUNT = 7;
    const UPDATE = 8;
    const VOUCHER_EXPIRE = 9;
    const NINETEEN = 10;

    const FREE = 11;
    const RIDIONLY = 12;
    const RIDISHOP = 13;
    const RIDISTORY = 14;
    const DAILY_BEST = 15;
    const SEVENTY_PERCENT = 16;
    const CASHBACK = 17;
    const CROWN = 18;
    const HIGHLIGHTED = 19;
    const JUSTTODAY = 20;

    const NEW = 21;
    const NOW = 22;
    const RECOMMENDATION = 23;
    const QUESTIONMARK = 24;
    const CHICKEN = 25;
    const CHICKEN2 = 26;
    const NOTI18_BAN = 27;
    const POINT = 28;
    const POINT_ALERT = 29;
    const EIGHTEEN_ROCK = 30;

    const NINE = 31;
    const WEEKEND = 32;
    const BL = 33;
    const TL = 34;
    const HD = 35;
    const DISCOUNT_B = 36;
    const DISCOUNT_P = 37;
    const DISCOUNT_Y = 38;
    const EVENT = 39;
    const UPDATE_P = 40;

    const DISCOUNT_W_P = 41;
    const DISCOUNT_W_Y = 42;
    const DISCOUNT_W_B = 43;
}
